bpo-32035: Update zipfile doc

srcecde opened the pull request:
Updated the cpython/Doc/library/zipfile.rst documentation.

<!-- issue-number: bpo-32035 -->
https://bugs.python.org/issue32035
<!-- /issue-number -->

srcecde committed:
issue #32035: Update zipfile doc

srcecde committed:
issue #32035: Update zipfile doc

terryjreedy committed:
bpo-30928: Update idlelib/NEWS.txt to 2017 Dec 3. (#4701)

Patches are in 3.7.0a3 even if this update is not.

terryjreedy committed:
bpo-32207: Improve tk event exception tracebacks in IDLE. (#4703)

When tk event handling is driven by IDLE's run loop, a confusing
and distracting queue.EMPTY traceback context is no longer added
to tk event exception tracebacks.  The traceback is now the same
as when event handling is driven by user code.  Patch based on
a suggestion by Serhiy Storchaka.

ericvsmith committed:
bpo-32214: Implement PEP 557: Data Classes (#4704)

terryjreedy committed:
bpo-30928: update idlelib/NEWS.txt. (#4706)

vadmium commented on Doc/library/zipfile.rst:
```diff
@@ -385,9 +385,9 @@ ZipFile Objects
-   name it will be given in the archive, or a :class:`ZipInfo` instance.  If it's
-   an instance, at least the filename, date, and time must be given.  If it's a
+   Write the string or bytes *data* to the archive; if it is a *str*, it should be encoded as UTF-8.
+   *zinfo_or_arcname* is either the name of the file in the archive or a :class:`ZipInfo` instance.
```
The doc string wording works better IMO; see <https://hg.python.org/cpython/rev/19196cce1431>. In particular “_bytes_ instance” makes more sense than “the bytes _data_”, and “it is encoded” is clearer; “it should be encoded” sounds like the responsibility of the caller.

Mariatta committed:
bpo-28791: Update macOS installer to use SQLite 3.21.0. (#4245)

ilevkivskyi committed:
[bpo-28556] Minor fixes for typing module (#4710)

ned-deily committed:
Enable SQLite JSON functions for macOS installer build (#4712)

nascheme committed:
bpo-19610: Warn if distutils is provided something other than a list to some fields (#4685)

* Rather than raise TypeError, warn and call list() on the value.

* Fix tests, revise NEWS and whatsnew text.

* Revise documentation, a string is okay as well.

* Ensure 'requires' and 'obsoletes' are real lists.

* Test that requires and obsoletes are turned to lists.

srcecde committed:
bpo-32035: Updated zipfile doc1

srcecde committed:
bpo-32035: Updated zipfile doc1

srcecde committed:
bpo-32035: Updated zipfile doc2

srcecde committed:
bpo-32035: Updated zipfile doc2

ned-deily committed:
bpo-31392: Update macOS installer to use OpenSSL 1.0.2m (#4715)

ned-deily committed:
Update macOS installer to XZ 5.2.3 (#4718)

ned-deily committed:
bpo-31380: Skip test_httpservers test_undecodable_file on macOS. (#4720)

The undecodable file name cannot be created on macOS APFS file systems.

JulienPalard committed:
Add a missing space in tkinter documentation. (GH-4692)

srcecde committed:
bpo-32035: Updated zipfile doc fix travis fail

srcecde committed:
bpo-32035: Updated zipfile doc fix travis fail

ned-deily committed:
Tidy NEWS entry.

ned-deily committed:
Update NEWS and pydoc topics.

ned-deily committed:
Bump to 3.7.0a3

vstinner committed:
bpo-32030: Simplify _PyCoreConfig_INIT macro (#4728)

* Simplify _PyCoreConfig_INIT, _PyMainInterpreterConfig_INIT,
  _PyPathConfig_INIT macros: no need to set fields to 0/NULL, it's
  redundant (the C language sets them to 0/NULL for us).
* Fix typo: pymain_run_statup() => pymain_run_startup()
* Remove a few XXX/TODO

ericvsmith committed:
Add dataclasses to CODEOWNERS. (GH-4730)

gvanrossum committed:
Add Guido and Ivan as owners for typing.py (#4729)

ned-deily committed:
Merge tag 'v3.7.0a3'

ned-deily committed:
Bump to 3.7.0a3+

vstinner committed:
bpo-32030: pass interp to _PyImport_Init() (#4736)

Remove also the initstr variable, unused since the commit
e69f0df45b709c25ac80617c41bbae16f56870fb pushed in 2012: "bpo-13959:
Re-implement imp.find_module() in Lib/imp.py"

vstinner committed:
bpo-32030: Add pymain_get_global_config() (#4735)

* Py_Main() now starts by reading Py_xxx configuration variables to
  only work on its own private structure, and then later writes back
  the configuration into these variables.
* Replace Py_GETENV() with pymain_get_env_var() which ignores empty
  variables.
* Add _PyCoreConfig.dump_refs
* Add _PyCoreConfig.malloc_stats
* _PyObject_DebugMallocStats() is now responsible to check if debug
  hooks are installed. The function returns 1 if stats were written,
  or 0 if the hooks are disabled. Mark _PyMem_PymallocEnabled() as
  static.

CuriousLearner committed:
bpo-25910: Fixes redirection from http to https (#4674)

vstinner committed:
bpo-32030: Add _PyImport_Fini2() (#4737)

PyImport_ExtendInittab() now uses PyMem_RawRealloc() rather than
PyMem_Realloc(). PyImport_ExtendInittab() can be called before
Py_Initialize() whereas only the PyMem_Raw allocator is supposed to
be used before Py_Initialize().

Add _PyImport_Fini2() to release the memory allocated by
PyImport_ExtendInittab() at exit. PyImport_ExtendInittab() now forces
the usage of the default raw allocator, to be able to release memory
in _PyImport_Fini2().

Don't export these functions anymore to be C API, only to
Py_BUILD_CORE:

* _PyExc_Fini()
* _PyImport_Fini()
* _PyGC_DumpShutdownStats()
* _PyGC_Fini()
* _PyType_Fini()
* _Py_HashRandomization_Fini()

Commit:
bpo-32233: Fix build with --with-system-libmpdec. (#4739)

datalogics-robb committed:
Modify configure to link with the compiler driver under HP-UX when not using gcc. (#2519)

ericvsmith committed:
Use raw strings for regex to avoid invalid escape sequences. (GH-4741)

zooba committed:
Allows non-critical upload steps to fail (#4742)

pablogsal committed:
Fix hardcoded value in test_os.py (#4744)

srcecde committed:
Merge branch 'fix-issue-32035' of https://github.com/srcecde/cpython into fix-issue-32035

serhiy-storchaka commented on Doc/library/zipfile.rst:
```diff
@@ -385,10 +385,11 @@ ZipFile Objects
-   name it will be given in the archive, or a :class:`ZipInfo` instance.  If it's
-   an instance, at least the filename, date, and time must be given.  If it's a
-   name, the date and time is set to the current date and time.
+   Write a file into archive. The content is *data* which may be either a *str* or a *bytes* instance;
```
Isn't this line too long?

`ZipFile.write()` writes a file into archive. `ZipFile.writestr()` writes a bytes data or a string.

serhiy-storchaka commented on Doc/library/zipfile.rst:
```diff
@@ -385,10 +385,11 @@ ZipFile Objects
-   an instance, at least the filename, date, and time must be given.  If it's a
-   name, the date and time is set to the current date and time.
+   Write a file into archive. The content is *data* which may be either a *str* or a *bytes* instance;
+   if it is a *str*, it is encoded as UTF-8.
```
I'm not sure this looks correct. `str` instance is not encoded. It is just a sequence of character, and how it is represented internally is an implementation detail.

srcecde replied:
Yes, you are right. Thanks for the update.

serhiy-storchaka commented on Doc/library/zipfile.rst:
```diff
@@ -385,10 +385,11 @@ ZipFile Objects
-   name, the date and time is set to the current date and time.
+   Write a file into archive. The content is *data* which may be either a *str* or a *bytes* instance;
+   if it is a *str*, it is encoded as UTF-8.
+   *zinfo_or_arcname* is either the name of the file in the archive or a :class:`ZipInfo` instance.
```
Please avoid unneeded rewrapping. Wrap only long lines that you have changed in any way.

srcecde committed:
Updated zipfile doc3

serhiy-storchaka commented on Doc/library/zipfile.rst:
```diff
@@ -385,10 +385,10 @@ ZipFile Objects
-   name it will be given in the archive, or a :class:`ZipInfo` instance.  If it's
-   an instance, at least the filename, date, and time must be given.  If it's a
-   name, the date and time is set to the current date and time.
+   `ZipFile.write()` writes a file into archive. `ZipFile.writestr()` writes a bytes data or a string.
```
Oh, I didn't mean that you should include my words verbatim. `ZipFile.write()` is documented in the description of `ZipFile.write()`.

The existing documentation is mostly good. The only problem is that the term "string" is used in wider meaning that includes byte strings, but this is not clear. Just document what actually does this method.

srcecde committed:
Updated zipfile doc4

Commit:
Merge branch 'fix-issue-32035' of https://github.com/srcecde/cpython into fix-issue-32035

Commit:
Updated zipfile doc4 travis fix
